File: Day297/busRoutes.java
class Solution {
    public int numBusesToDestination(int[][] routes, int source, int target) {
        if (source == target) {
            return 0;
        }
        if (routes.length == 1) {
            return -1;
        }
        Set<Integer> visited = new HashSet<>();
        Queue<Integer> queue = new LinkedList<>();
        Map<Integer, List<Integer>> map = new HashMap<>();
        int answer = 0;
        for (int i=0; i<routes.length; i++) {
            for (int j=0; j<routes[i].length; j++) {
                if (!map.containsKey(routes[i][j])) {
                    map.put(routes[i][j], new ArrayList<>());
                }
                map.get(routes[i][j]).add(i);
            }
        }
        queue.offer(source);
        while (!queue.isEmpty()) {
            int size = queue.size();
            answer++;
            for (int i=0; i<size; i++) {
                int temp = queue.poll();
                List<Integer> bus = map.get(temp);
                for (int v: bus) {
                    if (visited.contains(v)) {
                        continue;
                    }
                    visited.add(v);
                    for (int j=0; j<routes[v].length; j++) {
                        if (routes[v][j] == target) {
                            return answer;
                        }
                        queue.offer(routes[v][j]);
                    }
                }
            }
        }
        return -1;
    }
}

File: Day297/busRoutes.go
func numBusesToDestination(routes [][]int, source int, target int) int {
    if source == target {
        return 0
    }
    if len(routes) == 1 {
        return -1
    }
    m := map[int][]int{}
    for i,r := range routes {
        for _,s := range r {
            if len(m[s]) == 0 {
                m[s] = []int{}
            }
            m[s] = append(m[s], i)
        } 
    }
    queue := [][]int{{source, 0}}
    visitedR := map[int]struct{}{}
    visitedS := map[int]struct{}{}
    for len(queue) != 0 {
        temp := queue[0]
        queue = queue[1:]
        if _, flag := visitedS[temp[0]]; flag {
            continue
        }
        visitedS[temp[0]] = struct{}{}
        if temp[0] == target {
            return temp[1]
        }
        arr := m[temp[0]]
        for _, n := range arr {
            if _, flag := visitedR[n]; flag {
                continue
            }
            for _, r := range routes[n] {
                queue= append(queue, []int{r, temp[1]+1})
            }
            visitedR[n] = struct{}{}
        }
    }
    return -1
}